Remove error return on !exists

stevedotmt opened the pull request:
On Linux, volume binding creates the directory on the host. This
conditional check on Windows was preventing Windows Containers from
creating the directory producing inconsistent behavior between
platforms.

**- What I did**
Removed the hard check if the directory exists on the host in mount config validation.

**- How I did it**
Changed the condition logic.

**- How to verify it**
Mount a volume to a directory that does not exist on the host. It should be created and not throw an error.

**- Description for the changelog**
Remove feature parity between Linux & Windows when mounting directory that doesn't already exist.

**- A picture of a cute animal (not mandatory but encouraged)**
[Cute gopher](https://i.pinimg.com/236x/d4/7d/f0/d47df0a11725b7bdb9bada6b6f501360--ground-squirrel-wild-animals.jpg)

Commit:
Remove error return on !exists

On Linux, volume binding creates the directory on the host. This
conditional check on Windows was preventing Windows Containers from
creating the directory producing inconsistent behavior between
platforms.

Signed-off-by: Steve Azzopardi <sazzopardi@gitlab.com>

steveazz commented:
@thaJeztah I have created a new PR from https://github.com/moby/moby/pull/35159 to try and fix the tests, but I am slightly confused by the orignal fix.

On `windows_parser.go` we have the following change

```diff
diff --git a/volume/mounts/windows_parser.go b/volume/mounts/windows_parser.go
index ac610440436a..2ba9f0daedab 100644
--- a/volume/mounts/windows_parser.go
+++ b/volume/mounts/windows_parser.go
@@ -251,10 +251,7 @@ func (p *windowsParser) validateMountConfigReg(mnt *mount.Mount, destRegex strin
 		if err != nil {
 			return &errMountConfig{mnt, err}
 		}
-		if !exists {
-			return &errMountConfig{mnt, errBindSourceDoesNotExist(mnt.Source)}
-		}
-		if !isdir {
+		if exists && !isdir {
 			return &errMountConfig{mnt, fmt.Errorf("source path must be a directory")}
 		}
```

But don't we have the [same check in for `linux_parser.go`](https://github.com/moby/moby/blob/master/volume/mounts/linux_parser.go#L84-L89) why do need to remove the check for windows and not for the linux parser for it to create the non existant directory?

Would love some guideince! 🙇

thaJeztah commented:
Hm, interesting; I'd have to dig a bit, but could it be that the original PR was actually changing the wrong API? (i.e. the `--mount` option should not accept non-existing source-paths, but the `-v` / `--volume` flag should 🤔)


@cpuguy83 you're more familiar with these; perhaps you can point faster to the right location for the change?

steveazz commented:
Ok so after setting up a windows environments on my Mac I tested the following and it does seem to be the error that this PR removes. But we also throw that error on `linux_parser.go` which the same logic. 

Can it be the directory is being created before these checks are happening?

![screen shot 2019-01-02 at 17 29 56](https://user-images.githubusercontent.com/9993446/50601570-d7dd0880-0eb4-11e9-9830-4602dfa65a7a.png)

thaJeztah commented:
Hm, good point; that would explain yes. Looking if this is where that's happening (but not clear yet where the difference is made between windows and linux 🤔 ); https://github.com/moby/moby/blob/b3e9f7b13b0f0c414fa6253e1f17a86b2cff68b5/volume/mounts/mounts.go#L144-L163

steveazz commented:
Ok, I think I understand what is happening now *(I think)*. Inside of `linux_parser.go`  https://github.com/moby/moby/blob/e78a3dca21ea9ceb8d7a5e0252e94ee7ad05eadd/volume/mounts/linux_parser.go#L84-L89 

When `-v` is being a passed we call `ParseMountRaw` which calls [`p.parseMountSpec(spec, false)`](https://github.com/moby/moby/blob/master/volume/mounts/linux_parser.go#L272) so it does not validate if the source directory exists since `validateBindSourceExists` is false. The only time it is validated is when we call [`ParseMountSpec`](https://github.com/moby/moby/blob/master/volume/mounts/linux_parser.go#L282)

I would imagine we want the same logic for windows is that correct for `ParseMountSpec` & `PaeseMountRaw`

thaJeztah commented:
> I would imagine we want the same logic for windows is that correct

Yes, the intent of the original PR was to make `-v /non/existent/dir:/container-path` work the same on Windows and Linux (i.e., `/non/existent/dir` being created on the host)

thaJeztah commented:
Right, yes, `--mount` should never create the host-directory.

jhowardmsft commented on Dockerfile.windows:
```diff
@@ -153,7 +153,7 @@
 
 # The number of build steps below are explicitly minimised to improve performance.
-FROM microsoft/windowsservercore
+FROM mcr.microsoft.com/windows/servercore:ltsc2019
```
No, no, no, no!!!! This is a common dockerfile for many different builds (including nightly internally). It is expected that "microsoft/windowsservercore:latest" exists when building this. Changing this from as-is will break RS1 CI and more. Please revert this change. Thanks :)

steveazz replied:
Hello @jhowardmsft thanks for the heads up, but this is form my WIP commit which is a `git add .` not to lose anything when I closing the day, I will revert it when I am back, sorry for the hassle caused.

Commit:
Add check for bind source path exists

Add a new check to validate if the source path of the volume binding
exists. This is a is not always validated since in `ParseMountRaw` we
don't want to validate this.

This brings feature parity between the Linux parse and Windows parser,
where a directory is created if the source path doesn't exists.

Signed-off-by: Steve Azzopardi <sazzopardi@gitlab.com>

Commit:
Move source must be dir err to func

For consistency, testtability and reuseability sake move the string to a
func that returns a formatted error.

Signed-off-by: Steve Azzopardi <sazzopardi@gitlab.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "create-dirctory-on-host-for-windows" git@github.com:steveazz/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354568008
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

steveazz commented:
@cpuguy83 @thaJeztah I think this is ready to review can one of you take a look, appreciate any feedback! 🙇

steveazz commented:
I see the following failures for [pwerpc](https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/12875/console) build, which I am not sure it is related. Is that expected to fail?

steveazz commented:
There are some legit test failures that I need to look at

Commit:
Add specific mock for TestValidateMount

Create a new mock specific for the test `TestValidateMount` which
returns `exist` to `false` only when it's equal to
`testSourcePathDoesNotExist` so that the validation is done correctly.

Signed-off-by: Steve Azzopardi <sazzopardi@gitlab.com>

steveazz commented:
Ok, the final failure is not related to this PR and I am not sure if it should be fixed or removed altogether.

Looking at the failure below:

```
12:40:48 --- FAIL: TestParseMountRawSplit (0.00s)
12:40:48     parser_test.go:385: case 0
12:40:48     parser_test.go:385: case 1
12:40:48     parser_test.go:385: case 2
12:40:48     parser_test.go:385: case 3
12:40:48     parser_test.go:385: case 4
12:40:48     parser_test.go:385: case 5
12:40:48     parser_test.go:385: case 6
12:40:48     parser_test.go:385: case 7
12:40:48     parser_test.go:385: case 8
12:40:48     parser_test.go:385: case 0
12:40:48     parser_test.go:385: case 1
12:40:48     parser_test.go:385: case 2
12:40:48     parser_test.go:385: case 3
12:40:48     parser_test.go:385: case 4
12:40:48     parser_test.go:385: case 5
12:40:48     parser_test.go:385: case 6
12:40:48     parser_test.go:385: case 7
12:40:48     parser_test.go:385: case 8
12:40:48     parser_test.go:385: case 9
12:40:48     parser_test.go:385: case 10
12:40:48     parser_test.go:385: case 11
12:40:48     parser_test.go:385: case 12
12:40:48     parser_test.go:389: Expected error, was nil, for spec c:\foo\bar:\\.\pipe\foo
12:40:48     parser_test.go:385: case 0
12:40:48     parser_test.go:385: case 1
12:40:48     parser_test.go:385: case 2
12:40:48     parser_test.go:385: case 3
12:40:48     parser_test.go:385: case 4
12:40:48     parser_test.go:385: case 5
12:40:48     parser_test.go:385: case 6
12:40:48     parser_test.go:385: case 7
12:40:48     parser_test.go:385: case 8
12:40:48     parser_test.go:385: case 9
12:40:48     parser_test.go:385: case 10
12:40:48     parser_test.go:385: case 11
```

Looking a bit closer, it's this specific [case](https://github.com/moby/moby/blob/master/volume/mounts/parser_test.go#L355) that is failing. After checking out master, the error that is returned is `&errors.errorString{s:"invalid volume specification: 'c:\\foo\\bar:\\\\.\\pipe\\foo': invalid mount config for type \"bind\": bind source path does not exist: c:\\foo\\bar"}` so it's not really the expected error, it should be that the spec is invalid since one of them is a pipe and the other is not. Since this PR fixes the fact that the directory is created, it's not failing since we might have another bug somewhere. 

Inside [windowsDetectMountType](https://github.com/moby/moby/blob/master/volume/mounts/windows_parser.go#L166-L174) we try and figure out what mount type it is, by taking the source, which in this case ends up being `bind` since the source if a directory not a named pipe. So we never hit the [following validation](https://github.com/moby/moby/blob/master/volume/mounts/windows_parser.go#L275-L294)

What is the best way to procced with this?

steveazz commented:
@cpuguy83 I hope I am not genearting noise for you, but mind if you review this one?

steveazz commented:
Thanks @cpuguy83 for the feedback 🙇 in https://github.com/moby/moby/pull/38465#issuecomment-452702606 I said why it's failing and how the test is incorrect in the first place, mind if you check if my finding is correct and what is the best way forward?

steveazz commented:
@cpuguy83 or @thaJeztah would you mind if you help me out on the last failing test? When I looked into this in https://github.com/moby/moby/pull/38465#issuecomment-452702606 it seems like we are testing the wrong thing that is why it is failing.

Would appreciate any help!

cpuguy83 commented:
@steveazz Yes, not ignoring. Sorry just havne't had a chance to be able to dig into these test cases.... which are plentiful.

steveazz commented:
No worries @cpuguy83 didn't mean to rush you, thank you for all the help so far 🙏

cpuguy83 commented:
@steveazz It looks like the test was passing for invalid reasons before the change.
I have a feeling that it should be invalid to mount a dir to a named pipe.

cpuguy83 committed:
Make parser tests a little easier to read and decipher for test fails

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
I updated the tests so at least it's easier to tell what is going on when it fails.

cpuguy83 commented:
Still need to deal with making sure dir -> named pipe fails (I'm assuming this should not be allowed... /cc @jhowardmsft)

tallandtree commented:
Hi, just wondering if and when this will be solved. Can anyone give an update?
@cpuguy83 or @steveazz or @thaJeztah ?

steveazz commented:
@tallandtree there is one comment from me in https://github.com/moby/moby/pull/38465#issuecomment-452702606 that I'd like guidance on so we get the CI green after that it should be ready for review.

thaJeztah commented:
Let me try getting this discussed  in the maintainers meeting / review session (I might have another look at it before that).

In the meantime; could you rebase this PR? We replaced our CI with a new Jenkins, so it would be helpful to have it up to date, and have a fresh run of CI

Commit:
Add check for pipe for bind volume

When the mount type is set to `bind` check that the target is not a
named pipe.

Signed-off-by: Steve Azzopardi <sazzopardi@gitlab.com>

steveazz commented:
Ok I've updated an update the fixes the failure for `TestParseMountRawSplit` with 60f63c84bb38be6a9be8514b949c906692e8a05f but now I see a new failure in https://ci.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-38465/runs/4/nodes/53/steps/234/log/ I will spend some more time this weekend investigating to see if I can get the pipeline green 🙇

steveazz commented:
@thaJeztah I managed to make the pipeline green 🎉 I had to refactor the initial implementation a bit to make it successful and not modify any existing logic. I'm not 100% sure about the code quality, I welcome any feedback!
